### ISSUE TITLE
EKS: External traffic may be routed through wrong ENI and dropped, when masquerading is enabled

### ISSUE THREAD

**Issue by qmonnet** [cilium/cilium#18179]
TL;DR: When masquerading is enabled, external traffic from some pods may be routed through the wrong ENI on EKS, resulting in drops. This is likely due to some missing IP rules, to tell the system to look into the correct routing table (the one associated to the ENI to which the pod is attached).

### Symptoms

When running Cilium:

- On EKS
- With a non-trivial number of pods (e.g. 300)
- With masquerading enabled (`enable-ipv4-masquerade`)

... then external traffic may be broken for some pods. The symptoms are the following:

- External traffic for _some_ pods (requests originating from those pods, sent to the Internet) is dropped
- This is not a policy issue (occurs even with no policy enabled)
- Traffic internal to the cluster works as expected

Investigation showed that, for the affected pods, the external traffic is routed through the wrong ENI. Because of this, AWS drops it. For example, a pod attached to `eth3` may send its external traffic to `eth0` instead, resulting in drops.

### Reproduction

1. Create an EKS cluster;
2. Creating a VPC with CIDR `10.2.0.0/16`;
3. Associating secondary CIDRs `10.3{7,8,9}.0.0/16` to the VPC;
4. Creating one subnet for each of the aforementioned secondary CIDRs;
5. Creating an EKS cluster having nodes placed on the aforementioned subnets;
6. Install Cilium 1.10.4 with masquerading enabled (`enable-ipv4-masquerade`);
7. Create a simple `nginx` deployment (for example) with 300 replicas and let it stabilise;
8. Create a pod running `giantswarm/tiny-tools` (for example), run `ping 1.1.1.1` and check that `ping` fails.
9. Additionally check using `tcpdump` that traffic is leaving through the wrong interface.

Some of the pods should be unable to reach external targets. In our example, we run `ping 1.1.1.1` on `100.37.55.102` (one of the affected pods). This pod is attached to `eth3`:

```
root@ip-100-37-214-226:/home/cilium# ip rule | grep 100.37.55.102
20:     from all to 100.37.55.102 lookup main
111:    from 100.37.55.102 to 10.2.0.0/16 lookup 13
111:    from 100.37.55.102 to 100.37.0.0/16 lookup 13
111:    from 100.37.55.102 to 100.38.0.0/15 lookup 13

root@ip-100-37-214-226:/home/cilium# ip route show table 13
default via 100.37.0.1 dev eth3 <———
100.37.0.1 dev eth3 scope link

root@ip-100-37-214-226:/home/cilium# tcpdump -i eth3 -vvv host 1.1.1.1 <———
tcpdump: listening on eth3, link-type EN10MB (Ethernet), capture size 262144 bytes
[blank]

root@ip-100-37-214-226:/home/cilium# tcpdump -i eth0 -vvv host 1.1.1.1 <———
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
16:20:43.746802 IP (tos 0x0, ttl 254, id 54822, offset 0, flags [DF], proto ICMP (1), length 84)
    ip-100-37-55-102.eu-west-1.compute.internal > one.one.one.one: ICMP echo request, id 11008, seq 0, length 64
16:20:44.746988 IP (tos 0x0, ttl 254, id 54899, offset 0, flags [DF], proto ICMP (1), length 84)
    ip-100-37-55-102.eu-west-1.compute.internal > one.one.one.one: ICMP echo request, id 11008, seq 1, length 64
16:20:45.747173 IP (tos 0x0, ttl 254, id 54901, offset 0, flags [DF], proto ICMP (1), length 84)
    ip-100-37-55-102.eu-west-1.compute.internal > one.one.one.one: ICMP echo request, id 11008, seq 2, length 64
16:20:46.747335 IP (tos 0x0, ttl 254, id 55038, offset 0, flags [DF], proto ICMP (1), length 84)
    ip-100-37-55-102.eu-west-1.compute.internal > one.one.one.one: ICMP echo request, id 11008, seq 3, length 64
```

### Investigation

In the example above, we observe that traffic is routed through `eth0` instead of `eth3`, resulting in AWS dropping the external traffic for the affected pods.

Let's look further at the IP and routing rules. The IP for the affected pod has IP rules of the shape `from <IP> to <CIDR> lookup <ENI index>`:

```
111:    from 100.37.55.102 to 10.2.0.0/16 lookup 13
111:    from 100.37.55.102 to 100.37.0.0/16 lookup 13
111:    from 100.37.55.102 to 100.38.0.0/15 lookup 13
```

This works well for internal traffic, but this means that external traffic, which is not targetted to one of the VPC, matches none of the above rules. It fall backs to the default routing table, where it matches the following:

```
default via 100.38.0.1 dev eth0 
```

And then is routed through `eth0`. So this means that there is no kernel bug here, but the issue is related to the IP rules configuration.

By contrast _if we disabled masquerading_, the set of rules would be different. We would have, for the IP of the pod, rules of the shape `from <IP> lookup <ENI index>`, without the `to <CIDR>` filter. In that case, external traffic matches against the rule and picks the index of the routing table associated to the correct ENI, and goes out from the expected ENI (to which the pod is attached).

Looking at the code, all these rules come from the `Configure()` function in Cilium's `pkg/datapath/linux/routing/routing.go`:

```go
	if info.Masquerade {
		// Lookup a VPC specific table for all traffic from an endpoint to the
		// CIDR configured for the VPC on which the endpoint has the IP on.
		for _, cidr := range info.IPv4CIDRs {
			if err := route.ReplaceRule(route.Rule{
				Priority: egressPriority,
				From:     &ipWithMask,
				To:       &cidr,
				Table:    tableID,
			}); err != nil {
				return fmt.Errorf("unable to install ip rule: %s", err)
			}
		}
	} else {
		// Lookup a VPC specific table for all traffic from an endpoint.
		if err := route.ReplaceRule(route.Rule{
			Priority: egressPriority,
			From:     &ipWithMask,
			Table:    tableID,
		}); err != nil {
			return fmt.Errorf("unable to install ip rule: %s", err)
		}
	}
```

This snippet does the following:

- Without masquerading, we go through the `else`, we create those rules of shape `from <IP> lookup <ENI index>`, and they are used for external traffic, and this works.
- With masquerading enabled, we go through the `if`, we only create the rules of shape `from <IP> to <CIDR> lookup <ENI index>` that are necessary for internal traffic, and not the rules of the other shape, and external traffic is broken.

### Possible fix

From the above analysis, it looks like the IP rules of shape `from <IP> lookup <ENI index>` should be created unconditionally, in addition to the `to <CIDR>` rules when masquerading is enabled.

We (@bmcustodio and I) did a brief attempt at adding them, by taking the instructions out of this `else` block above. However, we could not observe the IP rules on the host, even though we validated that the code was supposed to run (adding a log line just above the snippet). We may have missed something and failed to have our code running correctly (?), or the addition of the IP rules failed for some reason, or the IP rules were removed for some reasons.

---

### Cilium Version

Discovered on cilium 1.10.4, but all supported versions are likely affected.

### Kernel Version

5.4.117-58.216.amzn2.x86_64

### Kubernetes Version

v1.20.7-eks-d88609

**Comment by austince** [comment]
Is there a workaround for this? Does setting the egress interface explicitly (e.g. `egressMasqueradeInterfaces: eth0` in Helm) help in any way?

**Comment by fazizsoltani** [comment]
Same here. In EKS we have 3 interfaces. Pods on eth2 and eth3 masqueraded with eth0 IP address instead of relative interface IP addresses.

**Comment by davidh83110** [comment]
> Same here. In EKS we have 3 interfaces. Pods on eth2 and eth3 masqueraded with eth0 IP address instead of relative interface IP addresses.

Same with you, and didn't find any workarounds. Any one has a possible quick solution?

**Comment by alex-hunt-materialize** [comment]
We haven't experienced issues with connectivity for outbound connections initiated from pods, but we did experience issues when pods on secondary ENIs with EIPs attached would try to respond to externally initiated connections to the EIP's address. https://github.com/cilium/cilium/issues/20554

The workaround was to inject an additional `ip rule` to use the local target rule table for all traffic from that pod. We did this with a daemonset that watched for pods labeled to get EIPs, looked for the cilium created rule table for the "native routing" traffic, and added an additional rule to use that rule table for all traffic, not just to destinations in the native routing CIDR.

https://github.com/MaterializeInc/k8s-eip-operator/tree/main/cilium_eip_no_masquerade_agent

**Comment by davidh83110** [comment]
> We haven't experienced issues with connectivity for outbound connections initiated from pods, but we did experience issues when pods on secondary ENIs with EIPs attached would try to respond to externally initiated connections to the EIP's address. #20554
> 
> The workaround was to inject an additional `ip rule` to use the local target rule table for all traffic from that pod. We did this with a daemonset that watched for pods labeled to get EIPs, looked for the cilium created rule table for the "native routing" traffic, and added an additional rule to use that rule table for all traffic, not just to destinations in the native routing CIDR.
> 
> https://github.com/MaterializeInc/k8s-eip-operator/tree/main/cilium_eip_no_masquerade_agent

Thanks, @alex-hunt-materialize. And yes, I did manually add the `ip rule` as a temporary workaround, thanks for your sharing, this is helpful for sure. However, my situation is simpler compared to yours, my ENIs are distributed in different subnets but have no EIP attached, the secondary ENI went to the wrong route table as "expected". So for me, I turns out to restrict the ENIs to the same subnet to solve the issue.

**Comment by sudeephb** [comment]
Hi, we have encountered this exact issue again. 
On a similar EKS setup with two VPCs and two ENIs, when the pod IP is assigned from secondary ENI, the rule to send external traffic through the right interface(rule 11 in the following instance) is not created for all the pods:

The following is a working pod, for which the external traffic goes through the right interface, and everything works.
```
[root@ip-100-65-12-164 /]# ip rule list | grep 100.65.9.179
20:     from all to 100.65.9.179 lookup main
111:    from 100.65.9.179 lookup 11
111:    from 100.65.9.179 to 10.0.0.0/16 lookup 11
111:    from 100.65.9.179 to 100.65.0.0/16 lookup 11
```

But, in some pods with IP assigned from secondary ENI:
```
[root@ip-100-65-12-164 /]# ip rule list | grep 100.65.2.39
20:     from all to 100.65.2.39 lookup main
111:    from 100.65.2.39 to 10.0.0.0/16 lookup 11
111:    from 100.65.2.39 to 100.65.0.0/16 lookup 11
```

This results in external traffic being routed through the default interface, which is then dropped.
The cilium version we're running is 1.18.6, installed via helm chart with similar configuration as above.

**Comment by pchaigno** [comment]
Please open a new issue, this has been long fixed so it's probably something different. You can refer to this issue in your new issue.
